model/project_ref.go

```diff
@@ -2451,20 +2459,25 @@ func (p *ProjectRef) GetProjectSetupCommands(opts apimodels.WorkstationSetupComm
 	return cmds, nil
 }
 
-func (p *ProjectRef) UpdateNextPeriodicBuild(definition string, nextRun time.Time) error {
-	for i, d := range p.PeriodicBuilds {
-		if d.ID == definition {
-			d.NextRunTime = nextRun
-			p.PeriodicBuilds[i] = d
-			break
-		}
+// UpdateNextPeriodicBuild updates the periodic build run time for the relevant collection.
+func UpdateNextPeriodicBuild(projectId, definition string, nextRun time.Time) error {
+	// Get the branch project on its own so we can determine where to update the run time.
+	branchProject, err := FindBranchProjectRef(projectId)
+	if err != nil {
+		return errors.Wrap(err, "finding branch project")
+	}
+	if branchProject == nil {
+		return errors.New("couldn't find project")
 	}
+
 	collection := ProjectRefCollection
-	idKey := ProjectRefIdKey
 	buildsKey := projectRefPeriodicBuildsKey
-	if p.UseRepoSettings() {
-		// if the periodic build is part of the repo then update there instead
-		repoRef, err := FindOneRepoRef(p.RepoRefId)
+	documentIdKey := ProjectRefIdKey
+	idToUpdate := branchProject.Id
+
+	// If periodic builds aren't defined for the project, see if it's part of the repo and update there instead.
+	if branchProject.PeriodicBuilds == nil && branchProject.UseRepoSettings() {
```

Comment:
Do repo refs ever run periodic builds, or only branch projects? Basically, is it possible for the branch project to have a periodic build defined, but the periodic build being updated in this function is defined at the repo level and not the branch project level?

Reply:
Repos only hold definitions that are inherited by branch projects, but we only inherit them if the branch doesn't have anything defined, i.e. the nil check here.